+Move longitudinal FreeSurfer outputs to longitudinal session folders…

mmilch01 opened the pull request:
See bc88380. This needs review before merging. 

+Move longitudinal FreeSurfer outputs to longitudinal session folders instead of symlinking
+remove target transform if it was created as symlink from a read-only source
+clean up large files in longitudinal diffusion output (NEEDS REVIEW)

Commit:
+Move longitudinal FreeSurfer outputs to longitudinal session folders instead of symlinking
+remove target transform if it was created as symlink from a read-only source
+clean up large files in longitudinal diffusion output (NEEDS REVIEW)

glasserm commented:
I don't think we need copies of much of anything from the cross-sectional ${Subject} folders persisting in the longitudinal ${Subject} folders.  I suspect these copies and symlinks were being done mainly to avoid needing to figure out exactly what files needed to be temporarily used. In the longitudinal folders, every single file (except I guess files that are time only and have no space), will need to either be in the template aligned T1w space or the new MNI space.  I would be better able to assist by reviewing the output files rather than the code.

Commit:
+Remove DWI files in raw space from longitudinal runs, addressing comment https://github.com/Washington-University/HCPpipelines/pull/374#discussion_r2651623956

mmilch01 commented on DiffusionPreprocessing/DiffPreprocPipeline.sh:
```diff
@@ -418,6 +418,21 @@ post_eddy_cmd=("${HCPPIPEDIR}/DiffusionPreprocessing/DiffPreprocPipeline_PostEdd
+        find "${StudyFolder}/${DWIName}" -type f -regex "${StudyFolder}/${DWIName}/${pattern}" -delete
+    done
+fi
+
```
No objections to this. Addressed in 7f4e3f6ac223c7056dae38e6c46ac0f54b3d20dc

glasserm reviewed (approved):
These look fine to me.

coalsont commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -650,6 +650,8 @@ if (( ! IsLongitudinal )); then
     ${FSLDIR}/bin/convert_xfm -omat ${WD}/fMRI2str_refinement-long.mat -concat "$T1wCross2LongXfm" ${WD}/fMRI2str_refinement.mat
+    #this needs for the following command to succeed if ${WD}/fMRI2str_refinement.mat is symlink pointing to read-only source.
+    rm -f ${WD}/fMRI2str_refinement.mat
     cp ${WD}/fMRI2str_refinement-long.mat ${WD}/fMRI2str_refinement.mat
```
Though, the manpage for GNU `cp` does say "if an existing destination file cannot be opened, remove it and try  again", suggesting that if the symlink were to a writable file on a writable filesystem, it would still be a symlink afterwards...so maybe a separate `rm -f` is the right call anyway.

coalsont commented on PostFreeSurfer/PostFreeSurferPipelineLongPrep.sh:
```diff
@@ -132,13 +132,14 @@ if (( ! TemplateProcessing )); then
+        rm -rf "$tp_folder_hcp"
+    fi
+    mv "$tp_folder_fslong" "$tp_folder_hcp"
+    if (( $? )); then 
```
We have `set -eu` and `source debug.shlib` in effect, so if `mv` exits nonzero, this conditional and message will never be reached, and should probably be removed.  If we really want a more custom message than the `debug.shlib` "while running/expanded call: mv ..." stuff, it needs to be done like this:

```
if ! mv ...
then
    log_Err_Abort "whatever"
fi
```

Commit:
+Address reviewer comment (remove unreachable code):
https://github.com/Washington-University/HCPpipelines/pull/374#discussion_r2663202055

mmilch01 committed:
+rebase to current master
+remove symlinks from cross-sectional sessions to template T1w folder, used by longitudinal FS
+remove 'fsaverage' symlink in template's T1s folder, created by FS
+added comments with example timepoint folder names in longitudinal Freesurfer and HCP structure.

Commit:
+rebased to current master
+Changed symlinks to relative where feasible. Left comment that the code needs attention in other places.

Commit:
+address
1) https://github.com/Washington-University/HCPpipelines/pull/374#discussion_r2729761345
2) https://github.com/Washington-University/HCPpipelines/pull/374#issuecomment-3802662386

coalsont commented on PostFreeSurfer/PostFreeSurferPipelineLongPrep.sh:
```diff
@@ -408,7 +411,7 @@ if (( TemplateProcessing ==  1 )); then
         fi
         mkdir -p "${AtlasSpaceFolder_timepoint}"
-        ln -sf "${AtlasSpaceFolder_template}/xfms" "${AtlasSpaceFolder_timepoint}/"
+        cp -r "${AtlasSpaceFolder_template}/xfms" "${AtlasSpaceFolder_timepoint}/"
```
The transforms folder shouldn't be large enough that symlinks are worth the trouble.

mmilch01 replied:
In the last few commits, I've already converted a number of absolute links to relative throughout multiple pipelines. There's currently hard constraint for longitudinal sessions to be in the same study folder as cross-sectional, so I don't see why not. This would be the only place that does hard copy instead of relative symlinks in similar situations.

mmilch01 replied:
I think I see your point. I'll let you know when the set of 'examples' is complete and ready for review, rather than posting them here one by one.

mmilch01 committed:
+fix tICAMakeCleanLongitudinalTemplate.sh arguments
+fix situation where non-existent "$MNIFolder/Results/$extractNameOut" folder breaks longitudinal processing.

mmilch01 committed:
+clean up

mmilch01 committed:
+removed parameters not needed in longitudinal mode. Tested to work.

coalsont commented on FreeSurfer/custom/conf2hires:
```diff
@@ -299,9 +300,11 @@ if($#MMode) then
   if(! -e $MM && $longitudinal) then
     set MMcross = ${SUBJECTS_DIR}/${tpNid}/mri/orig/${MMode}raw.mgz
     cd orig
+    # symlinks review: relative, OK
```
`$MMCross` looks absolute, should this comment be here?

But, we probably don't intend to change much about how freesurfer internal stuff works anyway.  We generally have to convert to nifti/gifti anyway, so symlinks to freesurfer-format files don't really matter to our pipelines.

mmilch01 committed:
Update TransmitBias/scripts/PseudoTransmit_IndividualAlignRawAndInitialMaps.sh

Co-authored-by: Tim Coalson <coalsont@users.noreply.github.com>

mmilch01 committed:
Update TransmitBias/scripts/PseudoTransmit_IndividualAlignRawAndInitialMaps.sh

Co-authored-by: Tim Coalson <coalsont@users.noreply.github.com>

mmilch01 committed:
Update fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh

Co-authored-by: Tim Coalson <coalsont@users.noreply.github.com>

mmilch01 committed:
Address:
https://github.com/Washington-University/HCPpipelines/pull/374/changes#r2968402173
https://github.com/Washington-University/HCPpipelines/pull/374/changes#r2968476086
https://github.com/Washington-University/HCPpipelines/pull/374/changes#r2968415790
https://github.com/Washington-University/HCPpipelines/pull/374/changes#r2968453047
https://github.com/Washington-University/HCPpipelines/pull/374/changes#r2968457379

Commit:
+address https://github.com/Washington-University/HCPpipelines/pull/374/changes#r2985155754

Commit:
+address https://github.com/Washington-University/HCPpipelines/pull/374#discussion_r2985126881

coalsont commented on Examples/Scripts/DiffusionPreprocessingBatch-long.sh:
```diff
@@ -115,18 +112,16 @@ for i in "${!Subjects[@]}"; do
 	    # all potentially need to be accounted for (which they are in Siemen's reported BWPPPE)
-	    EchoSpacingSec=0.00078
-	  
-	    PEdir=2 #Use 1 for Left-Right Phase Encoding, 2 for Anterior-Posterior
```
We don't need orphaned comment sections.  For things that longitudinal never needs, please delete the parameters and the comment blocks that explain the (now-removed) settings.

mmilch01 committed:
+remove comments for the unused parameters, addressing https://github.com/Washington-University/HCPpipelines/pull/374/changes#r3012658630

coalsont commented:
It looks like there is a conflict from adding longitudinal `-conf2hires` to master since this branch was created.  Please figure out a way to make github happy about it (manual rebase with conflict resolution edit and force push, or see what the github "resolve conflicts" button does).

mmilch01 committed:
+fix to allow merge into master (update with conf2hires option added on master branch)

mmilch01 commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```suggestion
            cp -rL "$fd" "$fMRIFolderLong/"
            chmod -R u+w "$fMRIFolderLong/"
```
ensure copied symlinks are dereferenced and resulting files are user-writable.

mmilch01 commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -689,14 +689,16 @@ else
         else
-            cp -r "$fd" "$fMRIFolderLong/"
+            cp -rL "$fd" "$fMRIFolderLong/"
+            chmod -R u+w "$fMRIFolderLong/"
```
```suggestion
```
Remove chmod to be consistent with the cross-sectional logic.

mmilch01 commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -689,14 +689,15 @@ else
         	continue
         else
-            cp -r "$fd" "$fMRIFolderLong/"
+            cp -rL "$fd" "$fMRIFolderLong/"
```
```suggestion
            cp -rL "$fd" "$fMRIFolderLong/"
            chmod -R u+w "$fMRIFolderLong/"
```

mmilch01 commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -689,14 +689,16 @@ else
         else
-            cp -r "$fd" "$fMRIFolderLong/"
+            cp -rL "$fd" "$fMRIFolderLong/"
+            chmod -R u+w "$fMRIFolderLong/"
```
```suggestion
            chmod -R +w "$fMRIFolderLong/"
```

coalsont replied:
FYI, you can make edits from the files tab by clicking the "..." at the top right of a file (when you have permissions to the repo the PR branch is on).

mmilch01 committed:
+rebae to current master
+fixed incorrect assignment of $AtlasSpaceFolder in longitudinal case and clearer disentanglement of AtlasSpaceFolder variables.

mmilch01 commented on FreeSurfer/LongitudinalFreeSurferPipeline.sh:
```diff
@@ -313,8 +314,8 @@ if (( end_stage > 0 )); then
     recon_all_cmd="recon-all.v6.hires"
-    recon_all_cmd+=" -sd ${LongDIR}"
-    recon_all_cmd+=" -long ${Session} ${TemplateID} -all"
+    recon_all_cmd+=" -sd ${TemplateT1wDir}"
+    recon_all_cmd+=" -long ${Session} ${TemplateID} -all -conf2hires"
```
```suggestion
    recon_all_cmd+=" -sd ${TemplateT1wDir}"
    recon_all_cmd+=" -long ${Session} ${TemplateID} -all"
```
roll back the -conf2hires option

mmilch01 committed:
roll back the conf2hires option

glasserm commented:
@mmilch01 @coalsont @myersm0 @mharms what is holding up merging this PR?

mharms commented:
I thought we had already handled all the longitudinal stuff.  Was this perhaps incorporated into a different PR?

glasserm commented:
We will merge this after diffusion testing in IntraDB and when conflicts are resolved.

coalsont commented on FreeSurfer/LongitudinalFreeSurferPipeline.sh:
```diff
@@ -319,8 +320,8 @@ if (( end_stage > 0 )); then
-    recon_all_cmd+=" -sd ${LongDIR}"
-    recon_all_cmd+=" -long ${Session} ${TemplateID} -all -conf2hires"
+    recon_all_cmd+=" -sd ${TemplateT1wDir}"
+    recon_all_cmd+=" -long ${Session} ${TemplateID} -all"
```
What is going on with `-conf2hires` now?

mmilch01 replied:
A glitch in manual merge I had to do. Fixed in 15bbd94b207796689f52dfadd878ded1fc3a0693

mmilch01 committed:
+ prevent mv from failing if target exists in case of pipeline re-run

mmilch01 commented on FreeSurfer/LongitudinalFreeSurferPipeline.sh:
```diff
@@ -319,8 +320,8 @@ if (( end_stage > 0 )); then
-    recon_all_cmd+=" -sd ${LongDIR}"
-    recon_all_cmd+=" -long ${Session} ${TemplateID} -all -conf2hires"
+    recon_all_cmd+=" -sd ${TemplateT1wDir}"
+    recon_all_cmd+=" -long ${Session} ${TemplateID} -all"
```
```suggestion
    recon_all_cmd+=" -long ${Session} ${TemplateID} -all -conf2hires"
```

mmilch01 committed:
Apply suggestion from @mmilch01

mmilch01 commented:
> @mmilch01 @coalsont @myersm0 @mharms what is holding up merging this PR?

I'm currently testing this rebased/merged PR on longitudinal human and cross-sectional NHP subject.

mmilch01 committed:
+Fix 'Human' if statement, wrong case that breaks longitudinal processing
+Fix the incorrect use of $Session (which in this context points to the whole longitudinal folder name), that breaks longitudinal processing

mmilch01 committed:
Merge branch 'longitudinal-file-management-patch' of https://github.com/Washington-University/HCPpipelines into longitudinal-file-management-patch

glasserm commented:
I don't see Misha's comment here, but he said: "Testing of the rebased PR was successful on NHP (up to fMRISurface) and longitudinal human (up to TransmitBias, excluding diffusion), with a couple of fixes for longitudinal mode. It is ready to merge in my view." @coalsont

mmilch01 committed:
+fix for unbound pipedirguessed variable

mmilch01 committed:
+fix initialization of transmitRes

mmilch01 commented:
> I don't see Misha's comment here, but he said: "Testing of the rebased PR was successful on NHP (up to fMRISurface) and longitudinal human (up to TransmitBias, excluding diffusion), with a couple of fixes for longitudinal mode. It is ready to merge in my view." @coalsont

I found two more issues with longitudinal TransmitBias (fixed), but it's still running. I'll post an update as soon as it finishes.

mmilch01 commented:
> I don't see Misha's comment here, but he said: "Testing of the rebased PR was successful on NHP (up to fMRISurface) and longitudinal human (up to TransmitBias, excluding diffusion), with a couple of fixes for longitudinal mode. It is ready to merge in my view." @coalsont

The test for TransmitBias were successful. MATLAB hung the first time due to unusual workload on the test machine. @coalsont, this is now fully tested and ready to merge.
